language/move-vm/runtime/src/loader.rs

```diff
@@ -681,6 +686,37 @@ impl Loader {
         check_natives_impl(module).map_err(|e| e.finish(Location::Module(module.self_id())))
     }
 
+    // We do not allow cyclic dependencies between modules, i.e., module A depends on module B
+    // while at the same time, B (transitively) depends on A.
+    fn verify_no_cyclic_module_dependencies(&self, module: &CompiledModule) -> VMResult<()> {
+        // A standard DFS algorithm that follows the dependency chain from the `source_module_id`
+        // to see whether the `dst_module_id` is in its dependencies. This function can only be
+        // called when the module is fully verified and the `module_cache` is properly prepared.
+        fn is_reachable_via_dependencies(
+            dst_module_id: &ModuleId,
+            src_module_id: &ModuleId,
+            loader: &Loader,
+        ) -> bool {
+            if dst_module_id == src_module_id {
+                return true;
+            }
+            let src_module = loader.get_module(src_module_id);
+            module_dependencies(src_module.module())
+                .iter()
+                .any(|dep| is_reachable_via_dependencies(dst_module_id, dep, loader))
```

Comment:
I would then maybe name these as next, so `|next| ...`

Just trying to make the DFS more apparent, but it's up to you